File: NQueens.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<List<String>> sols = new ArrayList<>();

    public List<List<String>> solveNQueens(int n) {
        List<String> board = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            board.add(new String(new char[n]).replace('\0', '.'));
        }
        solve(board, 0);
        return sols;
    }

    private boolean isSafe(List<String> board, int row, int col) {
        int n = board.size();
        for (int i = 0; i < n; i++) {
            if (board.get(i).charAt(col) == 'Q') return false;
            if (row - i >= 0 && col - i >= 0 && 
            board.get(row - i).charAt(col - i) == 'Q')
             return false;
            if (row - i >= 0 && col + i < n &&
             board.get(row - i).charAt(col + i) == 'Q') 
             return false;
        }
        return true;
    }

    private void solve(List<String> board, int row) {
        if (row == board.size()) {
            sols.add(new ArrayList<>(board));
            return;
        }

        for (int col = 0; col < board.size(); col++) {
            if (isSafe(board, row, col)) {
                board.set(row, board.get(row).substring(0, col) + 'Q' + 
                board.get(row).substring(col + 1));
                solve(board, row + 1);
                board.set(row, board.get(row).substring(0, col) + '.' + 
                board.get(row).substring(col + 1));
            }
        }
    }
}

File: NQueens.cpp
class Solution {
public:
 vector<vector<string>> ans;  //2D array to store ans
    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n, string(n, '.')); //empty board
        rec(board, 0);
        return ans;
    }
    bool safe(vector<string>& board, int row, int col){
        int n = size(board);
        for(int i=0; i<n; i++){
     //3 cases to check where to place the Qn so no killing happens
     if(board[i][col] == 'Q') 
     return false; //Check if Queen in same col
     if(row-i>=0 && col-i>=0 && board[row-i][col-i] == 'Q') 
    return false; //diagnolly left up in backward direction: 45 deg
     if(row-i>=0 && col+i<n && board[row-i][col+i] == 'Q') 
     return false; //diagnolly right up in forward direction: 135 deg

        }
        return true; // For all other possiblities
    }
    void rec(vector<string>& board, int row){
        int l = size(board);
        if(row==l){ //Reached last row, we hv a solution
        ans.push_back(board);
         return;
        }
        for(int col=0; col<l; col++){
            if(safe(board, row, col)){
                board[row][col] = 'Q'; //Placed on a valid cell
                rec(board, row+1);  //move to next row
                //Backtracking to get all possible ways
                board[row][col] = '.';
            }
        }
    }
};
